feat: Handle ANSI codes in test failure message

rluvaton opened the pull request:
Fixes #27 

## Todo
- [ ] add tests
- [ ] cleanup
- [x] Somehow make the output use the code font

Before:

<img width="1282" alt="image" src="https://github.com/connor4312/nodejs-testing/assets/16746759/a6be78d8-13e7-43d9-b592-1494390f6d34">

After:
<img width="749" alt="image" src="https://github.com/connor4312/nodejs-testing/assets/16746759/cd042bc5-568f-4b5e-b990-b8d16400ccf0">

rluvaton commented on src/runner.ts:
```diff
@@ -145,10 +146,20 @@ export class TestRunner {
+                    } else if (doesTextContainAnsiCodes(asText)) {
+                      const markdownString = new vscode.MarkdownString();
+                      markdownString.supportHtml = true;
+                      markdownString.isTrusted = true;
```
Need to verify what are the implications here as it been sanitized any way

rluvaton commented:
Fix conflicts

connor4312 commented:
Thanks! The loss of monospace is unfortunate, I wonder if we can bring it back in markdown or whether a change in VS Code would be needed for that...

connor4312 commented on src/runner.ts:
```diff
@@ -219,10 +220,19 @@ export class TestRunner implements vscode.Disposable {
+                  } else if (doesTextContainAnsiCodes(asText)) {
+                    const markdownString = new vscode.MarkdownString();
+                    markdownString.supportHtml = true;
+                    markdownString.isTrusted = true;
```
I don't think this text needs to be / should be marked as trusted since it's derived from the test runner output

rluvaton replied:
If it's not marked as trusted we can't color the output

https://github.com/microsoft/vscode/blob/6d2920473c6f13759c978dd89104c4270a83422d/src/vs/base/browser/markdownRenderer.ts#L308

connor4312 commented on src/ansi-converter.ts:
```diff
@@ -0,0 +1,100 @@
+  let hexColor: string;
+
+  // Modify the common colors to be nicer
+  // TODO - should somehow support the theme colors
```
[As of last week](https://github.com/microsoft/vscode/commit/126ded71ed6a2b74f142e3b6362ed2121046fb46) (currently only on VS Code Insiders) you can use the --vscode theme color variables here, including the [integrated terminal colors](https://code.visualstudio.com/api/references/theme-color#integrated-terminal-colors).

This would only cause issues if the terminal background color is configured to be a different type than the panel or peek backgrounds, but this is not common and using these variables is probably the best we can do

rluvaton replied:
I tried to make it work for hours when making the PR, because this is rendered as markdown and not HTML I can't use the variables

If we could have HTML output that would be great

connor4312 replied:
That's a permalink to a very old commit, the latest is https://github.com/microsoft/vscode/blob/97c8d7e7ccf9028bba26ea47d284bbce93e8a1e8/src/vs/base/browser/markdownRenderer.ts#L382

connor4312 commented on src/ansi-converter.ts:
```diff
@@ -0,0 +1,100 @@
+
+  // According to VS Code source code, only span supports style
+  // https://github.com/microsoft/vscode/blob/6d2920473c6f13759c978dd89104c4270a83422d/src/vs/base/browser/markdownRenderer.ts#L301
+  const html = parsed.map((span) => `<span style="${getSupportedCSSFromAnsiSpan(span)}">${span.text.replaceAll('\n', '<br/>')}</span>`).join("");
```
We should make sure to escape HTML found in the text if this isn't done by the ansicolor library

rluvaton replied:
The ansicolor does not escape html in the text, but html is sanitized anyway on vscode side

connor4312 reviewed (commented):
Direction looks good, some commons. And it seems my procrastination in reviewing this was useful ;)

rluvaton commented:
Regarding the trusted HTML and escaping, looking at vscode source code it handle the sanitize and escaping and protect from XSS

connor4312 commented:
Yea, but the test output could still include HTML (e.g. tests for a library that does anything with HTML) which should be rendered as text and not interpreted as HTML

rluvaton commented:
Thanks, will add test but making the output with monospace worked + using the css variables, thanks

also, great catch with the HTML that should be escaped

fixed the comments but still need to add tests, updated the image in the pr description to show the updated output

connor4312 commented on src/ansi-converter.ts:
```diff
@@ -0,0 +1,103 @@
+  const colorName = color.bright ? colorNameAsBright[color.name as keyof typeof colorNameAsBright] : color.name;
+
+  return `var(${colorNameToCSSVariableName[colorName as keyof typeof colorNameToCSSVariableName]})`;
+}
```
Overall this looks good. It looks like ansicolor [does not support truecolor](https://github.com/xpl/ansicolor/issues/17) yet so this handles the relevant cases for the moment

connor4312 commented:
Hi, thanks again for the PR.

I actually got another request for this on VS Code. Since this is a fairly common problem and the solution is (as demonstrated) kind of hard to get right for extensions, I have made a change in VS Code to respect and colorize ANSI sequences automatically https://github.com/microsoft/vscode/pull/207852

Again, I really appreciate you taking the time to contribute this! I'm sorry it didn't end up getting this way.

rluvaton commented:
Thank you! It's even better
